### PR TITLE
[sgd] Replaced class Resources in sgd with `use_gpu`

### PR DESCRIPTION
**Pull request by jichan3751**
## What do these changes do?
Removes duplicate class Resources in python/ray/experimental/sgd/pytorch/utils.py and replace with class Resources in /python/ray/tune/trial.py.

## Linter

- [x] I've run `scripts/format.sh` to lint the changes in this PR.

### CODE REVIEW

File: python/ray/experimental/sgd/pytorch/pytorch_trainer.py [hunk 1]
```diff
@@ -65,8 +66,7 @@ def __init__(self,
         self.optimizer_timer = utils.TimerStat(window_size=1)
 
         if resources_per_replica is None:
-            resources_per_replica = utils.Resources(
-                num_cpus=1, num_gpus=0, resources={})
+            resources_per_replica = Resources(cpu=1, gpu=0)
 
         if backend == "auto":
             backend = "nccl" if resources_per_replica.num_gpus > 0 else "gloo"
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/pytorch_trainer.py]: i think this needs to change

Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_trainer.py]: everywhere using `num_gpus` -> `gpu`

Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_trainer.py]: actually, hm maybe we should keep it and instead extend `Resources` to use `num_cpus` and `num_gpus`.

File: python/ray/experimental/sgd/pytorch/__init__.py [hunk 2]
```diff
@@ -3,6 +3,6 @@
 from __future__ import print_function
 
 from ray.experimental.sgd.pytorch.pytorch_trainer import PyTorchTrainer
-from ray.experimental.sgd.pytorch.utils import Resources
+from ray.tune.trial import Resources
```
Comment (richardliaw) [inline python/ray/experimental/sgd/pytorch/__init__.py]: and maybe move `Resources` into `ray.tune.resources`?

**Comment by richardliaw** [comment]
OK actually I got it; this change is fine so far. Let's make the SGD API simpler in this PR, and I think the work you've done so far is in progress towards that goal. 

Can you implement the following:

```python
class PyTorchTrainer(object):
    ...
    def __init__(self,
                 model_creator,
                 data_creator,
                 optimizer_creator=utils.sgd_mse_optimizer,
                 config=None,
                 num_replicas=1,
                 use_gpu=False,  # Make this easier
                 batch_size=16,
                 backend="auto"):

    @classmethod
    def default_resource_request(cls, resource):
        # currently wrong syntax
        return Resources(
            cpu=0,
            gpu=0,
            extra_cpu=cf["num_workers"],
            extra_gpu=int(use_gpu)* cf["num_workers"])
```

File: python/ray/experimental/sgd/pytorch/pytorch_trainer.py (hunk 1 above)
Comment (richardliaw) [reply python/ray/experimental/sgd/pytorch/pytorch_trainer.py]: `int(use_gpu)`

**Comment by jichan3751** [comment]
In order to I think I need to know what extra_gpu , extra_cpus mean in class Resources

File: python/ray/tune/trial.py [hunk 3]
```diff
@@ -300,7 +162,7 @@ def __init__(self,
                 if resources:
                     raise ValueError(
                         "Resources for {} have been automatically set to {} "
-                        "by its `default_resource_request()` method. Please "
+                    "by its `default_resource_request()` method. Please "
```
Comment (richardliaw) [inline python/ray/tune/trial.py]: (revert?)

**Comment by richardliaw** [comment]
> In order to apply 'use_gpu' and 'default_resource_request' I think I need to know what extra_gpu , extra_cpus mean in class Resources.
Ah ok, `extra_gpus` and `extra_cpus` are the resources that will be needed to run the training, but are not going to be used by the head process. For example, say I have 4 processes holding model replicas for distributed training. I also have 1 process that does coordination. The coordination process may be the main process, and it will launch the for 4 replica processes. Here, I would want 1 CPU and 0 GPU for the coordinator, and its 4 child processes I want 1 CPU and 1 GPU.

This would correspond to `Resources(cpu=1, gpu=0, extra_cpus=4, extra_gpus=4)`.

> Also, does removing resource_per_replica mean that user cannot specify to multiple gpus for each replica?

Yes, I think that's ok for now.

**Comment by jichan3751** [comment]
Can we try to merge this PR only with changing 'Resources' in experimental.sgd and moving Resources to new file ray.tune.resources first, and work on the default_resource_request, and int(use_gpu) with pytorch trainable class with seperate PR?

**Review by richardliaw — APPROVED** [review]

**Pull request merged**